Make ResourceTest restore workspace description #406

HeikoKlare opened the pull request:
Several realizations of a `ResourceTest` perform changes to the workspace description, e.g., in terms of changing the auto-build settings, without resetting them afterwards. And even if they do, the functionality is redundantly present at different places within the source code.
This induces dependencies between test executions, as the behavior of a subsequent test case execution depends on the modifications of the workspace description by previous ones.

The proposed changes do the following: 
- Store the workspace description before executing a `ResourceTest` and restore it during teardown
- Provide only one central method enabling and disabling autobuild and setting the build order and remove all redundant methods or in-place realizations
- Remove manual restorations of autobuild settings after test execution, as this is subsumed by `ResourceTest` performing `setup` and `teardown`
- Properly wait for background jobs completion in `BuilderTest` to avoid race conditions because of changed runtime behavior due to resetting workspace description after executing other test cases

I am not sure whether the missing restoration of the workspace settings after a test run leads to build failures right now (might be a cause for some of the random failures). But locally executing certain extracts of available tests, such as a single class or package, can currently lead to failures due to missing test isolation.

Contributes to #406.
Fixes #408.

jukzi commented:
@laubi can you please review the workflow changes?

laeubi commented:
> @Laubi can you please review the workflow changes?

I would assume they are only for testing purpose and should be reverted before merging this.

jukzi reviewed (commented):
the changes in *.java LGTM

jukzi commented:
> I would assume they are only for testing purpose and should be reverted before merging this.

@HeikoKlare please revert or explain the workflow changes.

HeikoKlare commented:
I rebased on the wrong branch, which pulled in the workflow change. Sorry for that!
I have removed the commit with the workflow changes. The other changes are the same after force push.
